internal/junit/reporter.go

```diff
@@ -25,6 +27,99 @@ func (r *Reporter) Add(t report.TestResult) {
 	r.TestResults = append(r.TestResults, t)
 }
 
+func parseJunitFiles(junits []report.Artifact) ([]TestSuites, error) {
+	var parsed []TestSuites
+	var errs []error
+	for _, ju := range junits {
+		if ju.Error != nil {
+			errs = append(errs, errors.New("failed to retried junit file"))
+			continue
+		}
+		ts, err := Parse(ju.Body)
+		if err != nil {
+			errs = append(errs, errors.New("failed to parse junit file"))
+			continue
+		}
+		parsed = append(parsed, ts)
+	}
+	if len(errs) > 0 {
+		return parsed, fmt.Errorf("%d errors occured while evaluating junit files", len(errs))
+	}
+	return parsed, nil
+}
+
+func reduceSuite(old TestSuite, new TestSuite) TestSuite {
+	testMap := map[string]int{}
+	for idx, tc := range old.TestCases {
+		key := fmt.Sprintf(`%s.%s`, tc.ClassName, tc.Name)
+		testMap[key] = idx
+	}
+
+	for _, tc := range new.TestCases {
+		key := fmt.Sprintf(`%s.%s`, tc.ClassName, tc.Name)
+		var idx int
+		var ok bool
+		if idx, ok = testMap[key]; !ok {
+			log.Warn().Str("test", key).Msg("Test has been found only in retries")
+			continue
+		}
+		old.TestCases[idx] = tc
+	}
+	return old
+}
+
+func reduceJunitFiles(junits []TestSuites) TestSuites {
+	suites := map[string]TestSuite{}
+
+	for _, junit := range junits {
+		for _, suite := range junit.TestSuites {
+			if _, ok := suites[suite.Name]; !ok {
+				suites[suite.Name] = suite
+				continue
+			}
+			suites[suite.Name] = reduceSuite(suites[suite.Name], suite)
+		}
+	}
+
+	output := TestSuites{}
+	for _, ts := range maps.Values(suites) {
+		ts.Tests = len(ts.TestCases)
+		ts.Errors = countErrors(ts.TestCases)
+		ts.Skipped = countSkipped(ts.TestCases)
+		output.TestSuites = append(output.TestSuites, ts)
+	}
+	return output
+}
+
+func countErrors(tcs []TestCase) int {
+	count := 0
+	for _, tc := range tcs {
+		if tc.Status == "error" {
+			count++
+		}
+	}
+	return count
+}
+func countSkipped(tcs []TestCase) int {
+	count := 0
+	for _, tc := range tcs {
+		if tc.Status == "skipped" {
+			count++
+		}
+	}
+	return count
+}
+
+func pickJunitFile(artifacts []report.Artifact) []report.Artifact {
```

Comment:
Nitpick: Just a slightly more convention based naming that I've encountered more often (also change to plural, since you aren't picking just one):
```suggestion
func filterJunitFiles(artifacts []report.Artifact) []report.Artifact {
```

or

```suggestion
func filterJunitArtifacts(artifacts []report.Artifact) []report.Artifact {
```